### ISSUE TITLE
can not upload a data file to use in benchmark

### ISSUE THREAD

**Issue by johnaohara** [Hyperfoil/Hyperfoil#16]
If you want to load data from a file, e.g. 

```
randomItem: 
  -  file:
```

There is currently no way to upload and reference in a distributed manor.  A full path name can be defined in the benchmark definition, but for a distributed benchmark, that file would need to be in the same location on all machines

**Comment by rvansa** [comment]
That's not entirely true. In this case, the file is loaded when `RandomItemStep$Builder.build()` is called. The `Benchmark` is stored on the server in the serialized form, so the data is there as well.
There's a glitch, however, so it's good that you pointed this out:

1. The benchmark can be uploaded in YAML form, so it's the server who loads the file when building the programmatic configuration. This results in different output if you upload the benchmark in YAML or as serialized object.
2. On server the benchmark is stored both in serialized form and as yaml (if that's available). The order of loading benchmarks is not set, and if YAML wins it will try to load the file on server even if client uploaded different version.

**Comment by rvansa** [comment]
Note that the agents never parse the benchmark; they always receive it as a complete set. So it's rather a matter of deciding whether the file should be loaded on client or controller.

1. Loading on client:
+ keeps `Benchmark` immutable and self-contained
- when uploading YAML we would have to send multi-part request with all the referenced files
- we couldn't store the YAML on server, or we would need to include all the files

2. Loading on server
+ keeps YAML upload simple
+ works the same regardless of where is the benchmark pushed from
- you'd need to push referenced files independently (I wouldn't make HF controller a generic file server)
- custom after-deserialization procedure to load all files
- the files can be mutated without control (which complicates things even more when the server caches the benchmarks)

**Comment by johnaohara** [comment]
How do I invoke the ControllerRestServer endpoint to upload the benchmark on a remote machine, and transmit the data.  I see that handlePostBenchmark deals with both YAML encoded and "application/java-serialized-object" content type, but in the serialized Benchmark object, I only see the following properties;

```
    protected final String name;
    protected final String originalSource;
    protected final Host[] agents;
    private final int threads;
    private final Ergonomics ergonomics;
    private final Map<String, Http> http;
    private final Http defaultHttp;
    private final Collection<Phase> phases;
    private final Map<String, Object> tags;
    private final long statisticsCollectionPeriod;
```
Where is the data file included in the serialized Benchmark definition?

It could handle a multi-part message, where we could transmit the yaml, and the data file(s) in one POST request.

**Comment by rvansa** [comment]
By `Benchmark` I mean the whole object graph it carries, the `phases` links through all the sequences and steps and `RandomItemStep` there contains the already parsed list (from the file).

**Comment by johnaohara** [comment]
OK, I hadn't dug that far down in the object graph, I can see now that RandomItemStep is serializable and it will serialize the String[] list of data

**Comment by rvansa** [comment]
Fixed in a67a2f46f26b83d0fcfa0d6639dc55a71b885f46
